### PR TITLE
[dotnet] Allow setting of PageDimensions and PageMargins in PrintOptions directly

### PR DESCRIPTION
**Pull request by shbenzer**
### **User description**
PrintOptions.cs was missing setters for PageMargins and PageDimensions.

### Description
added setters in PrintOptions.cs for PageMargins and PageDimensions.

### Motivation and Context
fixes issue

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

___

### **PR Type**
Bug fix

___

### **Description**
- Added missing setters for `PageDimensions` and `PageMargins` in the `PrintOptions` class to allow modification of these properties.
- This change addresses an issue where these properties were previously read-only, limiting their usability.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>PrintOptions.cs</strong><dd><code>Add missing setters for PageDimensions and PageMargins</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/PrintOptions.cs

<li>Added a setter for <code>PageDimensions</code> property.<br> <li> Added a setter for <code>PageMargins</code> property.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14593/files#diff-b7252358e015a773e2d453e794cf8c614fcd330985e0ddf4fe3202392f508e2c">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: dotnet/src/webdriver/PrintOptions.cs
Comment (nvborisenko) [inline dotnet/src/webdriver/PrintOptions.cs:1]: Please don't forget to adjust inline xml documentation for the changed properties. E.g. before: `Gets ...`, after `Gets or sets ...`

File: dotnet/src/webdriver/PrintOptions.cs [hunk 1]
```diff
@@ -106,6 +106,7 @@ public bool ShrinkToFit
         public PageSize PageDimensions
         {
             get { return pageSize; }
+            set { pageSize = value; }
```
Comment (nvborisenko) [inline dotnet/src/webdriver/PrintOptions.cs]: If we really want to allow set the value, then we need to add basic validation (for `null`).

File: dotnet/src/webdriver/PrintOptions.cs
Comment (shbenzer) [reply dotnet/src/webdriver/PrintOptions.cs:1]: Good catch, thank you!

File: dotnet/src/webdriver/PrintOptions.cs (hunk 1 above)
Comment (nvborisenko) [reply dotnet/src/webdriver/PrintOptions.cs]: Or allow it to be `null`? Then we should allow `null` for other properties, and then in code consider everywhere that properties might be null to avoid `NullReferenceExceptioon`.

From my point of view allowing it to be `null` is good idea, it represents an optional parameter according specification.

**Review by nvborisenko — CHANGES_REQUESTED** [review]
Allow properties to be `null` or not. Nullable properties is very good idea (see inline comment).

File: dotnet/src/webdriver/PrintOptions.cs [hunk 2]
```diff
@@ -199,6 +201,11 @@ internal Dictionary<string, object> ToDictionary()
                 toReturn["shrinkToFit"] = this.shrinkToFit;
             }
 
+            if (this.pageSize == null)
```
Comment (nvborisenko) [inline dotnet/src/webdriver/PrintOptions.cs]: Now it is better, which makes unnecessary to keep `this.pageSize` field initialized: https://github.com/SeleniumHQ/selenium/blob/4c2491faa3207cce1e9890bac16e98dfc7e5ee2e/dotnet/src/webdriver/PrintOptions.cs#L55

Comment (nvborisenko) [reply dotnet/src/webdriver/PrintOptions.cs]: Ah not, it is behavior breaking change. Please keep it as is.

File: dotnet/src/webdriver/PrintOptions.cs [hunk 3]
```diff
@@ -199,6 +201,11 @@ internal Dictionary<string, object> ToDictionary()
                 toReturn["shrinkToFit"] = this.shrinkToFit;
             }
 
+            if (this.pageSize == null) 
+            { 
+                this.pageSize = new PageSize(); 
+            }
+            
             if (this.pageSize.Height != DefaultPageHeight || this.pageSize.Width != DefaultPageWidth)
```
Comment (nvborisenko) [inline dotnet/src/webdriver/PrintOptions.cs]: It can be optimized to avoid allocation: `if (this.pageSize != null && (this.pageSize.Height != DefaultPageHeight || this.pageSize.Width != DefaultPageWidth))`

**Comment by nvborisenko** [comment]
Good progress here, I am going to checksum (using CI). If pass, the I vote for removing `DefaultPageHeight` value on client side, because I hope it should be handled oon server side.

**Comment by nvborisenko** [comment]
We don't have test for `Print` command. @shbenzer do you want to add new tests?

**Comment by shbenzer** [comment]
@nvborisenko sure, I wrote the documentation examples - they should easily port over.

**Comment by shbenzer** [comment]
Done

File: dotnet/test/common/PrintOptionsTest.cs
Comment (nvborisenko) [inline dotnet/test/common/PrintOptionsTest.cs:1]: Absolutely incorrect:
- We should use NUnit test framework
- There is no asserts
- There is no `Print` command invocation

I propose to take java tests as a reference to test `Print` command (hoping it exists :) )

Comment (nvborisenko) [inline dotnet/test/common/PrintOptionsTest.cs:1]: Thank you, I will help to adopt it to follow principles

Comment (nvborisenko) [reply dotnet/test/common/PrintOptionsTest.cs:1]: BTW do you use AI to write code?

**Comment by nvborisenko** [comment]
I looked at java implementation, it stores default values on client side. So let's be in sync.
- Add check for `null` when setting value (do not allow private value to be null)
- Add tests that value cannot be null
- Be in sync with java implementation when serializing object to json

**Comment by shbenzer** [comment]
> * Add check for `null` when setting value (do not allow private value to be null)

Do we want to throw an error when setting as null, or do we want to just set the default in that case and let it continue?

**Comment by nvborisenko** [comment]
> > * Add check for `null` when setting value (do not allow private value to be null)
> 
> Do we want to throw an error when setting as null, or do we want to just set the default in that case and let it continue?

Definitely throw argument null exception.

**Review by nvborisenko — APPROVED** [review]
Looks good, thank you

**Comment by shbenzer** [comment]
@nvborisenko Could you explain the code changes you made?

Particularly PageSizeDictionary()s changes, I can see the logic in the changes to the main sets, as well as refining the test example to just test construction instead of construction inside of the print command.

**Comment by nvborisenko** [comment]
Only for you..

1.
image
The same and simpler.

2.
image
In this case value is `double` (I don't remember), which is never can be `null`.

3.
image
`Assert` is correct, not `assert` which even could not be compiled. And the goal of this test is that we cannot set property to `null`, we don't test Print command.

**Comment by shbenzer** [comment]
Lol thank you

Those all make sense to me
